src/libraries/System.Private.Uri/src/System/PercentEncodingHelper.cs

```diff
@@ -0,0 +1,171 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Text;
+
+namespace System
+{
+    internal static class PercentEncodingHelper
+    {
+        public static unsafe int UnescapePercentEncodedUTF8Sequence(char* input, int length, ref ValueStringBuilder dest, bool isQuery, bool iriParsing)
+        {
+            // As an optimization, this method should only be called after the first character is known to be a part of a non-ascii UTF8 sequence
+            Debug.Assert(length >= 3);
+            Debug.Assert(input[0] == '%');
```

Comment:
Nit: This (and the two following asserts) are only valid if the input was originally a string. If the input could ever be a `ROS<char>`, this code must be audited to be resilient to the fact that the ROS contents could mutate while the method is executing. Here, "audited" means that the code would be allowed to produce garbage output given mutated input, but it shouldn't buffer overrun or go into an infinite loop or anything like that.

I don't think this is a problem right now since all the callers appear to work only with string instances, but this assumption should be called out somewhere as a code comment.